lib/khttp/ktest/mocks.go

```diff
@@ -0,0 +1,145 @@
+// +build !release
+
+package ktest
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	rpcAstore "github.com/enfabrica/enkit/astore/rpc/astore"
+	"github.com/enfabrica/enkit/astore/server/astore"
+	"github.com/enfabrica/enkit/lib/srand"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+	"log"
+	"math/rand"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+)
+
+type MinioDescriptor struct {
+	Port uint16
+}
+
+//RunMinioServer will spin up a minio serer using the local docker daemon
+//it also returns a func to call that will close and destroy the running image
+//the port and network bind are determined by docker and returned
+func RunMinioServer() (MinioDescriptor, func() error, error) {
+	_ = "gcloud beta emulators datastore start --no-store-on-disk"
+	return MinioDescriptor{}, nil, nil
+}
+
+type EmulatedDatastoreDescriptor struct {
+	Addr *net.TCPAddr
+}
+
+func RunEmulatedDatastore() (*EmulatedDatastoreDescriptor, KillAbleProcess, error) {
+	emulatorAddr, err := AllocatePort()
+	if err != nil {
+		return nil, nil, err
+	}
+	cmd := exec.Command("gcloud",
+		"beta", "emulators", "datastore", "start",
+		"--no-store-on-disk",
+		fmt.Sprintf("--host-port=127.0.0.1:%d", emulatorAddr.Port),
+		"--quiet")
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	outputStdErrPipe, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	err = cmd.Start()
+	killFunc := []func(){
+		func() {
+			fmt.Println("killing emulator datastore")
+			if err := cmd.Process.Kill(); err != nil {
+				log.Fatalln(fmt.Sprintf("error killing process %d", cmd.Process.Pid))
+			}
+		}}
+	if err != nil {
+		return nil, killFunc, err
+	}
+	datastoreBooted := make(chan bool)
+	//TODO concatenate stdout and stderr?
+	//the datastore emulator writes all logs to the error channel for some reason
+	scannerErr := bufio.NewScanner(outputStdErrPipe)
+	//TODO add timeout on starting, maybe 3 seconds?
```

Comment:
This should be really easy and important to do thanks to your choice of using a channel below. I would consider addressing it immediately. Something like this around line 78 would work?

```
select {
            case ret := <-datastoreBooted:
                return &EmulatorDescriptor{...}
            case <-time.After(10 * time.Second):
                kill...
                return failed to start error.
            }
```

Reply:
Implemented